Look into test failures

The `--parallel` mode of the tests is uncovering some more heisentests.

Known flaky tests:

**admin test**

```
ERROR: test_registry_list_for_defaults (openforms.prefill.tests.test_admin.PrefillConfigTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.10.9/x64/lib/python3.10/unittest/case.py", line 59, in testPartExecutor
    yield
  File "/opt/hostedtoolcache/Python/3.10.9/x64/lib/python3.10/unittest/case.py", line 591, in run
    self._callTestMethod(testMethod)
  File "/opt/hostedtoolcache/Python/3.10.9/x64/lib/python3.10/unittest/case.py", line 549, in _callTestMethod
    method()
  File "/home/runner/work/open-forms/open-forms/src/openforms/prefill/tests/test_admin.py", line 30, in test_registry_list_for_defaults
    self.assertGreater(len(response.form["default_company_plugin"].options), 1)
AttributeError: 'Text' object has no attribute 'options'
```

From Slack:

> actually, it sounds more to me like it's thread safety problems in dynamically altering the form field widget/type in `PrefillConfigAdmin`

**Excel file generation**

Very low occurrence, but some bytes are different sometimes, causing test failures?

**Unknown**

Some failures are unknown and eaten by the `--parallel` mode, for example here: https://github.com/open-formulieren/open-forms/actions/runs/4042602805/jobs/6950504659

```
"""
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.10.9/x64/lib/python3.10/multiprocessing/pool.py", line 125, in worker
    result = (True, func(*args, **kwds))
  File "/opt/hostedtoolcache/Python/3.10.9/x64/lib/python3.10/site-packages/django/test/runner.py", line 347, in _run_subsuite
    result = runner.run(subsuite)
  File "/opt/hostedtoolcache/Python/3.10.9/x64/lib/python3.10/site-packages/django/test/runner.py", line 294, in run
    test(result)
  File "/opt/hostedtoolcache/Python/3.10.9/x64/lib/python3.10/unittest/suite.py", line 84, in __call__
    return self.run(*args, **kwds)
  File "/opt/hostedtoolcache/Python/3.10.9/x64/lib/python3.10/unittest/suite.py", line 122, in run
    test(result)
  File "/opt/hostedtoolcache/Python/3.10.9/x64/lib/python3.10/site-packages/django/test/testcases.py", line 245, in __call__
    self._setup_and_call(result)
  File "/opt/hostedtoolcache/Python/3.10.9/x64/lib/python3.10/site-packages/django/test/testcases.py", line 281, in _setup_and_call
    super().__call__(result)
  File "/opt/hostedtoolcache/Python/3.10.9/x64/lib/python3.10/unittest/case.py", line 650, in __call__
    return self.run(*args, **kwds)
  File "/opt/hostedtoolcache/Python/3.10.9/x64/lib/python3.10/unittest/case.py", line 599, in run
    self._feedErrorsToResult(result, outcome.errors)
  File "/opt/hostedtoolcache/Python/3.10.9/x64/lib/python3.10/unittest/case.py", line 513, in _feedErrorsToResult
    result.addSubTest(test.test_case, test, exc_info)
  File "/opt/hostedtoolcache/Python/3.10.9/x64/lib/python3.10/site-packages/django/test/runner.py", line 251, in addSubTest
    self.check_subtest_picklable(test, subtest)
  File "/opt/hostedtoolcache/Python/3.10.9/x64/lib/python3.10/site-packages/django/test/runner.py", line 209, in check_subtest_picklable
    self._confirm_picklable(subtest)
  File "/opt/hostedtoolcache/Python/3.10.9/x64/lib/python3.10/site-packages/django/test/runner.py", line 142, in _confirm_picklable
    pickle.loads(pickle.dumps(obj))
AttributeError: Can't pickle local object '_get_target.<locals>.<lambda>'
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "src/manage.py", line 24, in <module>
    execute_from_command_line(sys.argv)
  File "/opt/hostedtoolcache/Python/3.10.9/x64/lib/python3.10/site-packages/django/core/management/__init__.py", line 419, in execute_from_command_line
    utility.execute()
  File "/opt/hostedtoolcache/Python/3.10.9/x64/lib/python3.10/site-packages/django/core/management/__init__.py", line 413, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/opt/hostedtoolcache/Python/3.10.9/x64/lib/python3.10/site-packages/django/core/management/commands/test.py", line 23, in run_from_argv
    super().run_from_argv(argv)
  File "/opt/hostedtoolcache/Python/3.10.9/x64/lib/python3.10/site-packages/django/core/management/base.py", line 354, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/opt/hostedtoolcache/Python/3.10.9/x64/lib/python3.10/site-packages/django/core/management/base.py", line 398, in execute
    output = self.handle(*args, **options)
  File "/opt/hostedtoolcache/Python/3.10.9/x64/lib/python3.10/site-packages/django/core/management/commands/test.py", line 55, in handle
    failures = test_runner.run_tests(test_labels)
  File "/opt/hostedtoolcache/Python/3.10.9/x64/lib/python3.10/site-packages/django/test/runner.py", line 729, in run_tests
    result = self.run_suite(suite)
  File "/opt/hostedtoolcache/Python/3.10.9/x64/lib/python3.10/site-packages/django/test/runner.py", line 670, in run_suite
    return runner.run(suite)
  File "/opt/hostedtoolcache/Python/3.10.9/x64/lib/python3.10/unittest/runner.py", line [184](https://github.com/open-formulieren/open-forms/actions/runs/4042602805/jobs/6950504659#step:12:185), in run
    test(result)
  File "/opt/hostedtoolcache/Python/3.10.9/x64/lib/python3.10/unittest/suite.py", line 84, in __call__
    return self.run(*args, **kwds)
  File "/opt/hostedtoolcache/Python/3.10.9/x64/lib/python3.10/site-packages/django/test/runner.py", line 411, in run
    subsuite_index, events = test_results.next(timeout=0.1)
  File "/opt/hostedtoolcache/Python/3.10.9/x64/lib/python3.10/multiprocessing/pool.py", line 873, in next
    raise value
  File "/opt/hostedtoolcache/Python/3.10.9/x64/lib/python3.10/multiprocessing/pool.py", line 125, in worker
    result = (True, func(*args, **kwds))
  File "/opt/hostedtoolcache/Python/3.10.9/x64/lib/python3.10/site-packages/django/test/runner.py", line 347, in _run_subsuite
    result = runner.run(subsuite)
  File "/opt/hostedtoolcache/Python/3.10.9/x64/lib/python3.10/site-packages/django/test/runner.py", line 294, in run
    test(result)
  File "/opt/hostedtoolcache/Python/3.10.9/x64/lib/python3.10/unittest/suite.py", line 84, in __call__
    return self.run(*args, **kwds)
  File "/opt/hostedtoolcache/Python/3.10.9/x64/lib/python3.10/unittest/suite.py", line 122, in run
    test(result)
  File "/opt/hostedtoolcache/Python/3.10.9/x64/lib/python3.10/site-packages/django/test/testcases.py", line 245, in __call__
    self._setup_and_call(result)
  File "/opt/hostedtoolcache/Python/3.10.9/x64/lib/python3.10/site-packages/django/test/testcases.py", line 281, in _setup_and_call
    super().__call__(result)
  File "/opt/hostedtoolcache/Python/3.10.9/x64/lib/python3.10/unittest/case.py", line 650, in __call__
    return self.run(*args, **kwds)
  File "/opt/hostedtoolcache/Python/3.10.9/x64/lib/python3.10/unittest/case.py", line 599, in run
    self._feedErrorsToResult(result, outcome.errors)
  File "/opt/hostedtoolcache/Python/3.10.9/x64/lib/python3.10/unittest/case.py", line 513, in _feedErrorsToResult
    result.addSubTest(test.test_case, test, exc_info)
  File "/opt/hostedtoolcache/Python/3.10.9/x64/lib/python3.10/site-packages/django/test/runner.py", line 251, in addSubTest
    self.check_subtest_picklable(test, subtest)
  File "/opt/hostedtoolcache/Python/3.10.9/x64/lib/python3.10/site-packages/django/test/runner.py", line [209](https://github.com/open-formulieren/open-forms/actions/runs/4042602805/jobs/6950504659#step:12:210), in check_subtest_picklable
    self._confirm_picklable(subtest)
  File "/opt/hostedtoolcache/Python/3.10.9/x64/lib/python3.10/site-packages/django/test/runner.py", line 142, in _confirm_picklable
    pickle.loads(pickle.dumps(obj))
AttributeError: Can't pickle local object '_get_target.<locals>.<lambda>'
```

This seems to happen with `subTest` at times?

Rare occurrence that the XLSX export failed locally:

```
FAIL: test_submission_with_email_backend_export_csv_xlsx (openforms.registrations.contrib.email.tests.test_backend.EmailBackendTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/bbt/.virtualenvs/open-forms/lib/python3.10/site-packages/django/test/utils.py", line 387, in inner
    return func(*args, **kwargs)
  File "/home/bbt/code/open-forms/src/openforms/registrations/contrib/email/tests/test_backend.py", line 454, in test_submission_with_email_backend_export_csv_xlsx
    self.assertEqual(xlsx_export[1], create_submission_export(qs).export("xlsx"))
AssertionError: b'PK\[29 chars]00\x98\x8c>V\x07AMb\x81\x00\x00\x00\xb1\x00\x0[14164 chars]\x00' != b'PK\[29 chars]00\x99\x8c>V\x07AMb\x81\x00\x00\x00\xb1\x00\x0[14164 chars]\x00'
```

I tried adding this test under the failing excel test:
```python
    def test_heisenbug(self):
        data = {"foo": "bar", "some_list": ["value1", "value2"]}

        submission = SubmissionFactory.create(form=self.form)
        submission_step = SubmissionStepFactory.create(
            submission=submission, form_step=self.fs, data=data
        )
        submission.completed_on = timezone.make_aware(datetime(2021, 1, 1, 12, 0, 0))
        submission.save()

        SubmissionFileAttachmentFactory.create(
            submission_step=submission_step,
            file_name="my-foo.bin",
            content_type="application/foo",
        )
        SubmissionFileAttachmentFactory.create(
            submission_step=submission_step,
            file_name="my-bar.txt",
            content_type="text/bar",
        )

        exported_data = []
        for _ in range(100):
            exported_data.append(create_submission_export(
                Submission.objects.filter(pk=submission.pk).select_related(
                    "auth_info"
                )
            ).export(AttachmentFormat.xlsx))

        for index, item in enumerate(exported_data[1:]):
            with self.subTest(index):
                self.assertEqual(item, exported_data[0])
```
This fails about 50% of the time when run on its own. When it fails:
1. as soon as one subtest fails, all the next ones fail too (but they don't always start from the same index)
2. The bytes shown in the assertion errors are always different between each run. For example, from three different failed runs:
```
AssertionError: b'PK\[29 chars]00\x8fT?V\x07AMb\x81\x00\x00\x00\xb1\x00\x00\x[14053 chars]\x00' != b'PK\[29 chars]00\x8eT?V\x07AMb\x81\x00\x00\x00\xb1\x00\x00\x[14053 chars]\x00'

AssertionError: b'PK\[29 chars]00\xc9T?V\x07AMb\x81\x00\x00\x00\xb1\x00\x00\x[14018 chars]\x00' != b'PK\[29 chars]00\xc8T?V\x07AMb\x81\x00\x00\x00\xb1\x00\x00\x[14018 chars]\x00'

AssertionError: b'PK\[29 chars]00\x02U?V\x07AMb\x81\x00\x00\x00\xb1\x00\x00\x[14080 chars]\x00' != b'PK\[29 chars]00\x01U?V\x07AMb\x81\x00\x00\x00\xb1\x00\x00\x[14080 chars]\x00'
```

--- a/src/openforms/prefill/tests/test_admin.py
+++ b/src/openforms/prefill/tests/test_admin.py
@@ -1,8 +1,11 @@
+from unittest.mock import patch
+
 from django.urls import reverse
 
 from django_webtest import WebTest
 
 from openforms.accounts.tests.factories import SuperUserFactory
+from openforms.config.models import GlobalConfiguration
 
 from ..models import PrefillConfig
 
@@ -13,7 +16,19 @@ class PrefillConfigTests(WebTest):
 
         self.assertEqual(str(instance), PrefillConfig._meta.verbose_name)
 
-    def test_registry_list_for_defaults(self):
+    @patch(
+        "openforms.plugins.registry.GlobalConfiguration.get_solo",
+        return_value=GlobalConfiguration(
+            plugin_configuration={
+                "registration": {
+                    "stufbg": {"enabled": True},
+                    "haalcentraal": {"enabled": True},
+                    "kvk-kvknumber": {"enabled": True},
+                },
+            }
+        ),
+    )
+    def test_registry_list_for_defaults(self, m_global_config):
         user = SuperUserFactory.create()
         (
             config,

--- a/src/openforms/registrations/contrib/email/tests/test_backend.py
+++ b/src/openforms/registrations/contrib/email/tests/test_backend.py
@@ -8,10 +8,15 @@ from django.urls import reverse
 from django.utils import timezone
 from django.utils.translation import ugettext_lazy as _
 
+import tablib
 from furl import furl
 
 from openforms.config.models import GlobalConfiguration
-from openforms.forms.tests.factories import FormFactory
+from openforms.forms.tests.factories import (
+    FormDefinitionFactory,
+    FormFactory,
+    FormStepFactory,
+)
 from openforms.payments.constants import PaymentStatus
 from openforms.payments.tests.factories import SubmissionPaymentFactory
 from openforms.submissions.exports import create_submission_export
@@ -38,14 +43,31 @@ from ..plugin import EmailRegistration
 class EmailBackendTests(HTMLAssertMixin, TestCase):
     @classmethod
     def setUpTestData(cls):
-        cls.form = FormFactory.create(
-            generate_minimal_setup=True,
+        fd = FormDefinitionFactory.create(
+            configuration={
+                "components": [
+                    {
+                        "key": "someField",
+                        "label": "Some Field",
+                        "type": "textfield",
+                    },
+                    {
+                        "key": "someList",
+                        "label": "Some list",
+                        "type": "textfield",
+                        "multiple": True,
+                    },
+                ],
+            }
+        )
+        form = FormFactory.create(
             name="MyName",
             internal_name="MyInternalName",
             registration_backend="email",
         )
-        cls.fs = cls.form.formstep_set.get()
-        cls.fd = cls.fs.form_definition
+        cls.fs = FormStepFactory.create(form=form, form_definition=fd)
+        cls.form = form
+        cls.fd = fd
 
     def test_submission_with_email_backend(self):
         submission = SubmissionFactory.from_components(
@@ -408,7 +430,7 @@ class EmailBackendTests(HTMLAssertMixin, TestCase):
             attachment_formats=[AttachmentFormat.csv, AttachmentFormat.xlsx],
         )
 
-        data = {"foo": "bar", "some_list": ["value1", "value2"]}
+        data = {"someField": "value0", "someList": ["value1", "value2"]}
 
         submission = SubmissionFactory.create(form=self.form)
         submission_step = SubmissionStepFactory.create(
@@ -421,11 +443,13 @@ class EmailBackendTests(HTMLAssertMixin, TestCase):
             submission_step=submission_step,
             file_name="my-foo.bin",
             content_type="application/foo",
+            form_key="attachment1",
         )
         SubmissionFileAttachmentFactory.create(
             submission_step=submission_step,
             file_name="my-bar.txt",
             content_type="text/bar",
+            form_key="attachment2",
         )
 
         email_submission = EmailRegistration("email")
@@ -451,7 +475,33 @@ class EmailBackendTests(HTMLAssertMixin, TestCase):
         self.assertEqual(
             xlsx_export[0], f"{submission.form.admin_name} - submission.xlsx"
         )
-        self.assertEqual(xlsx_export[1], create_submission_export(qs).export("xlsx"))
+
+        data = tablib.Dataset()
+        binary_excel = xlsx_export[1]
+        data.load(binary_excel)
+
+        self.assertEqual(
+            data.headers,
+            [
+                "Formuliernaam",
+                "Inzendingdatum",
+                "someField",
+                "someList",
+                "attachment1",
+                "attachment2",
+            ],
+        )
+        self.assertEqual(
+            list(data._data[0]),
+            [
+                "MyInternalName",
+                datetime(2021, 1, 1, 12, 0),
+                "value0",
+                "['value1', 'value2']",
+                None,
+                None,
+            ],
+        )
         self.assertEqual(
             xlsx_export[2],
             "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",

--- a/src/openforms/registrations/contrib/stuf_zds/tests/test_failure_modes.py
+++ b/src/openforms/registrations/contrib/stuf_zds/tests/test_failure_modes.py
@@ -6,6 +6,7 @@ import requests_mock
 from freezegun import freeze_time
 from privates.test import temp_private_root
 
+from openforms.config.models import GlobalConfiguration
 from openforms.submissions.constants import RegistrationStatuses
 from openforms.submissions.tests.factories import SubmissionFactory
 from stuf.stuf_zds.models import StufZDSConfig
@@ -37,6 +38,16 @@ class PartialRegistrationFailureTests(StUFZDSTestBase):
         cls.config_patcher = patch(
             "openforms.registrations.contrib.stuf_zds.plugin.StufZDSConfig.get_solo",
             return_value=StufZDSConfig(service=cls.service),
+        )
+        cls.general_config_patcher = patch(
+            "openforms.plugins.plugin.GlobalConfiguration.get_solo",
+            return_value=GlobalConfiguration(
+                plugin_configuration={
+                    "registration": {
+                        "stuf_zds": {"enabled": True},
+                    },
+                }
+            ),
         )
 
         # set up a simple form to track the partial result storing state
@@ -71,6 +82,9 @@ class PartialRegistrationFailureTests(StUFZDSTestBase):
 
         self.config_patcher.start()
         self.addCleanup(self.config_patcher.stop)
+
+        self.general_config_patcher.start()
+        self.addCleanup(self.general_config_patcher.stop)
 
         self.requests_mock = requests_mock.Mocker()
         self.requests_mock.start()

--- a/src/openforms/registrations/contrib/stuf_zds/tests/test_failure_modes.py
+++ b/src/openforms/registrations/contrib/stuf_zds/tests/test_failure_modes.py
@@ -95,9 +95,8 @@ class PartialRegistrationFailureTests(StUFZDSTestBase):
             additional_matcher=match_text("zakLk01"),
         )
 
+        register_submission(self.submission.id)
         with self.subTest("Initial registration fails"):
-            register_submission(self.submission.id)
-
             self.submission.refresh_from_db()
             self.assertEqual(
                 self.submission.registration_status, RegistrationStatuses.failed
@@ -172,9 +171,8 @@ class PartialRegistrationFailureTests(StUFZDSTestBase):
             additional_matcher=match_text("genereerDocumentIdentificatie_Di02"),
         )
 
+        register_submission(self.submission.id)
         with self.subTest("Document id generation fails"):
-            register_submission(self.submission.id)
-
             self.submission.refresh_from_db()
             self.assertEqual(
                 self.submission.registration_status, RegistrationStatuses.failed
@@ -268,9 +266,8 @@ class PartialRegistrationFailureTests(StUFZDSTestBase):
             additional_matcher=match_text("edcLk01"),
         )
 
+        register_submission(self.submission.id)
         with self.subTest("Document id generation fails"):
-            register_submission(self.submission.id)
-
             self.submission.refresh_from_db()
             self.assertEqual(
                 self.submission.registration_status, RegistrationStatuses.failed

